component/attr_cache/attr_cache.go

```diff
@@ -501,46 +478,63 @@ func (ac *AttrCache) RenameDir(options internal.RenameDirOptions) error {
 	err := ac.NextComponent().RenameDir(options)
 
 	if err == nil {
+		ac.cacheLock.Lock()
+		defer ac.cacheLock.Unlock()
+
+		// TLDR: Dst is guaranteed to be non-existent or empty.
+		// Note: We do not need to invalidate children of Dst due to the logic in our FUSE connector, see comments there,
+		// but it is always safer to double check than not.
+		ac.invalidateDirectory(options.Dst)
+
+		// if attr_cache is tracking directories, validate this rename
 		if ac.cacheDirs {
-			// renameCachedDirectory may cache the parent directory
-			// so lock the cache for writing
-			ac.cacheLock.Lock()
-			defer ac.cacheLock.Unlock()
-			err = ac.renameCachedDirectory(options.Src, options.Dst, currentTime)
+			// First, check if the destination directory already exists
+			if ac.pathExistsInCache(options.Dst) {
+				return os.ErrExist
+			}
+		}
+
+		// get the source directory
+		srcItem, getErr := ac.cacheMap.get(options.Src)
+		if getErr != nil || !srcItem.exists() {
+			log.Err("AttrCache::RenameDir : %s source not found", options.Src)
+			if ac.cacheDirs {
+				return syscall.ENOENT
+			}
 		} else {
-			ac.cacheLock.RLock()
-			defer ac.cacheLock.RUnlock()
-			ac.deleteDirectory(options.Src, currentTime)
-			// TLDR: Dst is guaranteed to be non-existent or empty.
-			// Note: We do not need to invalidate children of Dst due to the logic in our FUSE connector, see comments there,
-			// but it is always safer to double check than not.
-			ac.invalidateDirectory(options.Dst)
+			// move everything over
+			srcDir := internal.TruncateDirName(options.Src)
+			dstDir := internal.TruncateDirName(options.Dst)
+			ac.moveAttrCachedItem(srcItem, srcDir, dstDir, currentTime)
 		}
 	}
 
 	return err
 }
 
-// CreateFile: Mark the file invalid
+// CreateFile: Cache a new entry for the file
 func (ac *AttrCache) CreateFile(options internal.CreateFileOptions) (*handlemap.Handle, error) {
 	log.Trace("AttrCache::CreateFile : %s", options.Name)
 	h, err := ac.NextComponent().CreateFile(options)
 
 	if err == nil {
+		currentTime := time.Now()
 		// TODO: the cache locks are used incorrectly here
 		// They routinely lock the cache for reading, but then write to it
 		ac.cacheLock.RLock()
 		defer ac.cacheLock.RUnlock()
 		if ac.cacheDirs {
 			// record that the parent directory tree contains at least one object
-			ac.markAncestorsInCloud(getParentDir(options.Name), time.Now())
+			ac.markAncestorsInCloud(getParentDir(options.Name), currentTime)
 		}
-		// TODO: we assume that the OS will call GetAttr after this.
-		// 		if it doesn't, will invalidating this entry cause problems?
-		toBeInvalid, getErr := ac.cacheMap.get(options.Name)
-		if getErr == nil {
-			toBeInvalid.invalidate()
+		newFileEntry, getErr := ac.cacheMap.get(options.Name)
```

Comment:
Another thought here, so before they would check that if there was an entry existing for the file they invalidate it, wheras we create it. Do we need to worry about an old cache item being in the cacheMap or can we be guaranteed that won't happen or won't cause an issue?

Reply:
Since our `insert` function uses a map for each directory's files, insert will replace any entry with the same path.
Here the code checks if the file exists, and if it does, it updates its size, mode, and timestamp.
As I'm re-reading this code, I think it would have been better to just replace the entry outright, and not update an old entry.
I'll add a separate suggestion comment here shortly for your review.